File: python/examples/12_choosing_a_variant.py
"""Choosing a variant: batching, device placement, and measured throughput.

The whole point of a batched library is that many small problems are solved in
one launch. This example measures that, and shows how to pick between the syev
variants for a given (n, batch) instead of guessing.

Timings here include the host round trip (NumPy in, NumPy out), which is what
you actually pay from Python.

Run with:  python 12_choosing_a_variant.py
"""

from __future__ import annotations

import numpy as np

import batchlas as bl

from _common import (
    batched_symmetric,
    eigenvalue_error,
    header,
    preferred_device,
    report,
    section,
    timed,
)


def main() -> None:
    header("12. Choosing a variant")
    device = preferred_device()
    report("device", device or "library default")

    section("One launch beats a Python loop")
    # Same arithmetic either way; the difference is launch overhead and occupancy.
    batch, n = 512, 16
    matrices = batched_symmetric(batch, n, seed=1)

    _, batched_time = timed(lambda: bl.syev_cta(matrices, compute_vectors=False, device=device))

    def one_at_a_time():
        return [bl.syev_cta(matrices[i], compute_vectors=False, device=device) for i in range(batch)]

    _, looped_time = timed(one_at_a_time, repeats=1)

    report(f"batched  ({batch} x {n}x{n})", f"{batched_time * 1e3:.2f} ms")
    report(f"looped   ({batch} calls)", f"{looped_time * 1e3:.2f} ms")
    report("speed-up", f"{looped_time / batched_time:.1f}x")

    section("How throughput scales with batch size")
    for batch_size in (1, 16, 128, 1024):
        matrices = batched_symmetric(batch_size, 16, seed=2)
        _, elapsed = timed(lambda: bl.syev_cta(matrices, compute_vectors=False, device=device))
        report(
            f"batch={batch_size:<5d}",
            f"{elapsed * 1e3:7.2f} ms total, {elapsed / batch_size * 1e6:7.1f} us per matrix",
        )

    section("Which syev variant wins at which size?")
    # syev_cta and syev_jacobi_cta only accept n <= 32; the blocked and two-stage
    # paths are built for larger n. Measure rather than assume.
    for n, batch_size in ((16, 512), (32, 256), (128, 32), (512, 4)):
        matrices = batched_symmetric(batch_size, n, seed=3)
        reference = np.linalg.eigvalsh(matrices)
        support = bl.syev_variant_support(matrices, device=device)
        candidates = ["syev"]
        if n <= 32 and support["cta"]:
            candidates += ["syev_cta", "syev_jacobi_cta"]
        if support["blocked"]:
            candidates.append("syev_blocked")
        if support["two_stage"]:
            candidates.append("syev_two_stage")

        report(f"n={n}, batch={batch_size}", "")
        for name in candidates:
            try:
                values, elapsed = timed(
                    lambda fn=name: getattr(bl, fn)(matrices, compute_vectors=False, device=device)
                )
                error = eigenvalue_error(values, reference)
                report(f"  {name:16s}", f"{elapsed * 1e3:8.2f} ms   max eigenvalue error {error:.2e}")
            except (RuntimeError, NotImplementedError) as exc:
                report(f"  {name:16s}", f"unavailable ({type(exc).__name__})")

    section("CPU versus GPU for the same call")
    # Small batches often do not justify a device transfer; large ones clearly do.
    # Note syev_cta and friends need a sub-group width of 32 and so are GPU-only;
    # the general syev driver runs on both.
    for batch_size in (1, 256):
        matrices = batched_symmetric(batch_size, 16, seed=4)
        line = []
        for target in ("cpu", "gpu"):
            try:
                _, elapsed = timed(lambda t=target: bl.syev(matrices, compute_vectors=False, device=t))
                line.append(f"{target}: {elapsed * 1e3:8.2f} ms")
            except (RuntimeError, NotImplementedError) as exc:
                line.append(f"{target}: unavailable ({type(exc).__name__})")
        report(f"batch={batch_size:<5d}", "   ".join(line))

    section("Skipping eigenvectors is a real saving")
    matrices = batched_symmetric(256, 32, seed=5)
    _, with_vectors = timed(lambda: bl.syev_cta(matrices, compute_vectors=True, device=device))
    _, values_only = timed(lambda: bl.syev_cta(matrices, compute_vectors=False, device=device))
    report("with eigenvectors", f"{with_vectors * 1e3:.2f} ms")
    report("values only", f"{values_only * 1e3:.2f} ms")

    section("float32 versus float64")
    matrices = batched_symmetric(512, 16, seed=6)
    for dtype in (np.float32, np.float64):
        typed = matrices.astype(dtype)
        _, elapsed = timed(lambda m=typed: bl.syev_cta(m, compute_vectors=False, device=device))
        report(f"{np.dtype(dtype).name}", f"{elapsed * 1e3:.2f} ms")


if __name__ == "__main__":
    main()

File: examples/cpp/12_choosing_a_variant.cc
// 12. Choosing a variant
//
// The measurement example. What batching actually buys, how throughput scales
// with batch size, which syev variant wins at which size, and two easy savings.
//
// Timings are wall-clock around a queue that is waited on, so they include
// launch overhead — which is the point: for small matrices that overhead *is*
// the cost, and amortising it across a batch is the whole reason this library
// exists. The numbers come from whatever machine you run on.

#include <algorithm>
#include <chrono>
#include <cstddef>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <vector>

#include <blas/linalg.hh>
#include <util/sycl-vector.hh>

#include "example_utils.hh"

using namespace batchlas;
using namespace examples;

namespace {

// Median wall-clock milliseconds of `reps` runs, after a warm-up.
template <typename F>
double time_ms(Queue& ctx, F&& f, int reps = 5) {
    f();
    ctx.wait();
    std::vector<double> samples;
    for (int r = 0; r < reps; ++r) {
        const auto t0 = std::chrono::steady_clock::now();
        f();
        ctx.wait();
        samples.push_back(std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - t0).count());
    }
    std::sort(samples.begin(), samples.end());
    return samples[samples.size() / 2];
}

std::string ms(double v, int places = 2) {
    std::ostringstream os;
    os << std::fixed << std::setprecision(places) << v << " ms";
    return os.str();
}

template <Backend B>
struct Example {
    static void run(Queue& ctx) {
        batching_section(ctx);
        scaling_section(ctx);
        variant_section(ctx);
        device_section(ctx);
        savings_section(ctx);
    }

    // One launch beats a loop
    //
    // The same work, submitted as one batched call versus one call per matrix.
    // For small matrices the per-launch overhead dominates completely — on a
    // CPU device, where that overhead is small, expect a modest gap.
    static void batching_section(Queue& ctx) {
        section("One launch beats a loop");

        const int n = 16, batch = 256;

        auto A = Matrix<double>::Random(n, n, true, batch, 42);
        UnifiedVector<double> w(static_cast<size_t>(n) * batch);
        UnifiedVector<std::byte> ws(
            syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower));
        ctx.wait();
        const double batched = time_ms(ctx, [&] {
            syev<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws.to_span());
        });

        auto One = Matrix<double>::Random(n, n, true, 1, 42);
        UnifiedVector<double> w1(n);
        UnifiedVector<std::byte> ws1(
            syev_buffer_size<B>(ctx, One.view(), w1.to_span(), JobType::NoEigenVectors, Uplo::Lower));
        ctx.wait();
        const double looped = time_ms(
            ctx,
            [&] {
                for (int i = 0; i < batch; ++i)
                    syev<B>(ctx, One.view(), w1.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws1.to_span());
            },
            3);

        print(std::to_string(batch) + " matrices of size " + std::to_string(n) + ", batched", ms(batched));
        print(std::to_string(batch) + " matrices of size " + std::to_string(n) + ", one at a time", ms(looped));
        std::ostringstream os;
        os << std::fixed << std::setprecision(1) << looped / std::max(batched, 1e-9) << "x";
        print("speed-up", os.str());
    }

    // How throughput scales with batch size
    //
    // Time per matrix falls as the batch grows, until the device is saturated
    // and it flattens out. Where that happens is what you want to know when
    // sizing work.
    static void scaling_section(Queue& ctx) {
        section("How throughput scales with batch size");

        const int n = 32;
        for (int batch : {1, 8, 64, 512}) {
            auto A = Matrix<double>::Random(n, n, true, batch, 42);
            UnifiedVector<double> w(static_cast<size_t>(n) * batch);
            UnifiedVector<std::byte> ws(
                syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower));
            ctx.wait();
            const double t = time_ms(ctx, [&] {
                syev<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws.to_span());
            });
            std::ostringstream os;
            os << ms(t) << " total, " << std::fixed << std::setprecision(1) << t * 1000.0 / batch
               << " us per matrix";
            print("batch " + std::to_string(batch), os.str());
        }
    }

    // Which syev variant wins at which size?
    //
    // syev picks for you. These are the variants it picks between, timed
    // directly so you can see the crossovers.
    //
    // Run with JobType::EigenVectors: syev_two_stage returns wrong eigenvalues
    // in values-only mode for n >= 32 (see the README), so a values-only
    // comparison would be timing a broken path.
    //
    // These routines overwrite A, so repeated timing runs operate on the
    // previous result — the same amount of work on different data, which is
    // what we are measuring.
    static void variant_section(Queue& ctx) {
        section("Which syev variant wins at which size?");

        if (!on_gpu(ctx)) {
            skip("variant comparison", "the alternatives to the vendor path are GPU only");
            return;
        }

        const int batch = 64;
        for (int n : {16, 32, 64, 128}) {
            auto run_one = [&](const char* name, auto&& call) {
                auto A = Matrix<double>::Random(n, n, true, batch, 7);
                UnifiedVector<double> w(static_cast<size_t>(n) * batch);
                ctx.wait();
                const double t = time_ms(ctx, [&] { call(A, w); }, 3);
                print("n=" + std::to_string(n) + " " + name, ms(t));
            };

            run_one("syev (dispatched)", [&](Matrix<double>& A, UnifiedVector<double>& w) {
                UnifiedVector<std::byte> ws(
                    syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower));
                syev<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower, ws.to_span());
            });

            if constexpr (has_cta_variants<B>) {
                if (n <= 32 && supports_cta(ctx)) {
                    run_one("syev_cta         ", [&](Matrix<double>& A, UnifiedVector<double>& w) {
                        UnifiedVector<std::byte> ws(syev_cta_buffer_size<B>(ctx, A.view(), JobType::EigenVectors));
                        syev_cta<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower, ws.to_span());
                    });
                }
            }

            run_one("syev_blocked     ", [&](Matrix<double>& A, UnifiedVector<double>& w) {
                UnifiedVector<std::byte> ws(
                    syev_blocked_buffer_size<B>(ctx, A.view(), JobType::EigenVectors, Uplo::Lower));
                syev_blocked<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower, ws.to_span());
            });

            run_one("syev_two_stage   ", [&](Matrix<double>& A, UnifiedVector<double>& w) {
                UnifiedVector<std::byte> ws(
                    syev_two_stage_buffer_size<B>(ctx, A.view(), JobType::EigenVectors, Uplo::Lower));
                syev_two_stage<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower, ws.to_span());
            });
        }
    }

    // CPU versus GPU
    //
    // The backend is fixed at compile time, so this cannot switch mid-run. What
    // it can do is report what this device manages — run the example again with
    // `cpu` for the other half of the comparison.
    static void device_section(Queue& ctx) {
        section("CPU versus GPU");

        print("device", ctx.device().get_name());
        print("compute units", ctx.device().get_property(DeviceProperty::MAX_COMPUTE_UNITS));
        print("max work-group size", ctx.device().get_property(DeviceProperty::MAX_WORK_GROUP_SIZE));

        const int n = 32, batch = 256;
        auto A = Matrix<double>::Random(n, n, true, batch, 42);
        UnifiedVector<double> w(static_cast<size_t>(n) * batch);
        UnifiedVector<std::byte> ws(
            syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower));
        ctx.wait();
        const double t = time_ms(ctx, [&] {
            syev<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws.to_span());
        });

        std::ostringstream os;
        os << ms(t) << " (" << std::fixed << std::setprecision(1) << t * 1000.0 / batch << " us each)";
        print(std::to_string(batch) + " eigenproblems of size " + std::to_string(n), os.str());
        std::cout << "run this example again with the `cpu` argument to compare\n";
    }

    // Two easy savings
    //
    // Ask for eigenvalues only when you do not need vectors, and reuse a
    // workspace across calls instead of sizing and allocating each time.
    static void savings_section(Queue& ctx) {
        section("Two easy savings");

        const int n = 64, batch = 64;
        auto A = Matrix<double>::Random(n, n, true, batch, 42);
        UnifiedVector<double> w(static_cast<size_t>(n) * batch);
        UnifiedVector<std::byte> ws_v(
            syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower));
        UnifiedVector<std::byte> ws_n(
            syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower));
        ctx.wait();

        const double with_vectors = time_ms(
            ctx,
            [&] { syev<B>(ctx, A.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower, ws_v.to_span()); }, 3);
        const double without = time_ms(
            ctx,
            [&] { syev<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws_n.to_span()); }, 3);

        print("with eigenvectors", ms(with_vectors));
        print("eigenvalues only", ms(without));
        print("workspace with vectors", ws_v.size());
        print("workspace without", ws_n.size());

        // Sizing and allocating inside a loop pays for a device allocation
        // every iteration.
        const int reps = 20;
        const double reused = time_ms(
            ctx,
            [&] {
                for (int i = 0; i < reps; ++i)
                    syev<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws_n.to_span());
            },
            3);
        const double reallocated = time_ms(
            ctx,
            [&] {
                for (int i = 0; i < reps; ++i) {
                    UnifiedVector<std::byte> tmp(
                        syev_buffer_size<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower));
                    syev<B>(ctx, A.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, tmp.to_span());
                }
            },
            3);

        print(std::to_string(reps) + " calls, workspace reused", ms(reused));
        print(std::to_string(reps) + " calls, workspace reallocated each time", ms(reallocated));
    }
};

}  // namespace

BATCHLAS_EXAMPLE_MAIN("12. Choosing a variant")
